File: sol/3362/3362.py
import heapq as hq

class Solution:
    def maxRemoval(self, nums: List[int], queries: List[List[int]]) -> int:
        n = len(nums)
        diff = [0] * (n + 1)
        cumsum = 0
        q = len(queries)
        queries.sort(key=lambda itv: itv[0])
        h = []
        qptr = 0
        count = 0
        for i in range(n):
            cumsum += diff[i]
            while qptr < q and queries[qptr][0] <= i:
                hq.heappush(h, -queries[qptr][1])
                qptr += 1
            while len(h) != 0 and cumsum < nums[i]:
                r = -hq.heappop(h)
                if r >= i:
                    cumsum += 1
                    diff[r + 1] -= 1
                    count += 1
            if cumsum < nums[i]:
                return -1
        return q - count

File: sol/3362/3362.cpp
struct Solution {
    int maxRemoval(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size(), q = queries.size(), cumsum = 0, qptr = 0, count = 0;
        vector<int> diff(n + 1, 0), h;
        ranges::sort(queries, [&](const auto &a, const auto &b){return a[0] < b[0];});
        for (int i = 0; i < n; ++i){
            cumsum += diff[i];
            while (qptr < q && queries[qptr][0] <= i){
                h.push_back(queries[qptr++][1]);
                ranges::push_heap(h);
            }
            while (!h.empty() && cumsum < nums[i]){
                ranges::pop_heap(h);
                int r = h.back();
                h.pop_back();
                if (r >= i){
                    ++cumsum;
                    --diff[r + 1];
                    ++count;
                }
            }
            if (cumsum < nums[i]){
                return -1;
            }
        }
        return q - count;
    }
};
